lib/gst/player/gstplayer-media-info.c

```diff
@@ -0,0 +1,649 @@
+}
+
+static GstPlayerStreamInfo*
+gst_player_video_info_copy (GstPlayerVideoInfo *ref)
+{
+  GstPlayerVideoInfo *ret;
+
+  ret = gst_player_video_info_new ();
+
+  ret->width = ref->width;
+  ret->height  = ref->height;
+  ret->framerate_num = ref->framerate_num;
+  ret->framerate_denom = ref->framerate_denom;
+  ret->par_num = ref->par_num;
+  ret->par_denom = ref->par_denom;
+  ret->bitrate = ref->bitrate;
+  ret->max_bitrate = ref->max_bitrate;
+
+  return (GstPlayerStreamInfo*) ret;
+}
+
+static GstPlayerStreamInfo*
+gst_player_audio_info_copy (GstPlayerAudioInfo *ref)
+{
+  GstPlayerAudioInfo *ret;
+
+  ret = gst_player_audio_info_new ();
+
+  ret->sample_rate = ref->sample_rate;
+  ret->channels = ref->channels;
+  ret->bitrate = ref->bitrate;
+  ret->max_bitrate = ref->max_bitrate;
+
+  if (ref->language)
+    ret->language = g_strdup (ref->language);
+
+  return (GstPlayerStreamInfo*) ret;
+}
+
+static GstPlayerStreamInfo*
+gst_player_subtitle_info_copy (GstPlayerSubtitleInfo *ref)
+{
+  GstPlayerSubtitleInfo *ret;
+
+  ret = gst_player_subtitle_info_new ();
+  if (ref->language)
+    ret->language = g_strdup (ref->language);
+
+  return (GstPlayerStreamInfo*) ret;
+}
+
+GstPlayerStreamInfo*
+gst_player_stream_info_copy (GstPlayerStreamInfo *ref)
+{
+  GstPlayerStreamInfo *info = NULL;
+
+  if (GST_IS_PLAYER_VIDEO_INFO(ref))
+    info = gst_player_video_info_copy ((GstPlayerVideoInfo*)ref);
+
+  if (GST_IS_PLAYER_AUDIO_INFO(ref))
+    info = gst_player_audio_info_copy ((GstPlayerAudioInfo*)ref);
+
+  if (GST_IS_PLAYER_SUBTITLE_INFO(ref))
+    info = gst_player_subtitle_info_copy ((GstPlayerSubtitleInfo*)ref);
+
+  info->stream_index = ref->stream_index;
+  if (ref->tags)
+    info->tags = gst_tag_list_copy (ref->tags);
+  if (ref->caps)
+    info->caps = gst_caps_copy (ref->caps);
+
+  return info;
+}
+
+GstPlayerMediaInfo*
+gst_player_media_info_copy (GstPlayerMediaInfo* ref)
+{
+  GList *l;
+  GstPlayerMediaInfo *info;
+
+  if (!ref)
+    return NULL;
+
+  info = gst_player_media_info_new (ref->uri);
+
+  for (l = ref->stream_list; l != NULL; l = l->next)
+    info->stream_list = g_list_append (info->stream_list,
+                  gst_player_stream_info_copy((GstPlayerStreamInfo*)l->data));
+
+  return info;
+}
+
+GstPlayerStreamInfo*
+gst_player_stream_info_new (gint stream_index, GType type)
+{
+  GstPlayerStreamInfo *info = NULL;
+
+  if (type == GST_TYPE_PLAYER_AUDIO_INFO)
+    info = (GstPlayerStreamInfo*) gst_player_audio_info_new ();
+
+  if (type == GST_TYPE_PLAYER_VIDEO_INFO)
+    info = (GstPlayerStreamInfo*) gst_player_video_info_new ();
+
+  if (type == GST_TYPE_PLAYER_SUBTITLE_INFO)
+    info = (GstPlayerStreamInfo*) gst_player_subtitle_info_new ();
+
+  info->stream_index = stream_index;
+
+  return info;
+}
+
+GstPlayerMediaInfo*
+gst_player_media_info_new (const gchar *uri)
+{
+  GstPlayerMediaInfo *info;
+
+  info = g_object_new (GST_TYPE_PLAYER_MEDIA_INFO, NULL);
+  if (uri)
+    info->uri = g_strdup (uri);
```

Comment:
I think this must be called with a valid URI, add a g_return_val_if_fail(uri != NULL, NULL) at the top :)